cpp/src/copying/get_element.cu

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/copying.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/dictionary/dictionary_column_view.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {
+
+struct get_element_functor {
+  
+  template<typename T>
+  std::enable_if_t<is_fixed_width<T>(), 
+  std::unique_ptr<scalar>> operator()(
+    column_view const& input,
+    size_type index,
+    cudaStream_t stream = 0,
+    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
+  {
+    auto s = make_fixed_width_scalar(data_type(type_to_id<T>()), stream, mr);
+
+    using ScalarType = cudf::experimental::scalar_type_t<T>;
+    auto typed_s = static_cast<ScalarType *>(s.get());
+    
+    auto device_s = get_scalar_device_view(*typed_s);
+    auto device_col = column_device_view::create(input, stream);
+    
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream), 
+                       thrust::make_constant_iterator(0), 1, 
+      [ device_s, d_col=*device_col, index ] __device__ (auto) mutable {
+        device_s.set_value(d_col.element<T>(index));
+        device_s.set_valid(d_col.is_valid(index));
+      });
+    return s;
+  }
+
+  template<typename T>
+  std::enable_if_t<std::is_same<T, string_view>::value, 
+  std::unique_ptr<scalar>> operator()(
+    column_view const& input,
+    size_type index,
+    cudaStream_t stream = 0,
+    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
+  {
+    auto device_col = column_device_view::create(input, stream);
+
+    rmm::device_scalar<string_view> temp_data;
+    rmm::device_scalar<bool> temp_valid;
+
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream), 
+                       thrust::make_constant_iterator(0), 1, 
+      [ 
+        buffer=temp_data.data(),
+        validity=temp_valid.data(),
+        d_col=*device_col,
+        index 
+      ] __device__ (auto) mutable {
+        *buffer = d_col.element<string_view>(index);
+        *validity = d_col.is_valid(index);
+      });
+
+    return std::make_unique<string_scalar>(
+      temp_data, temp_valid.value(stream), stream, mr);
```

Comment:
This is copying the `device_scalar`s. Isn't there a ctor that takes ownership?

Reply:
Practically it’s possible for the validity scalar but not for `string_view` which is non-owning. That data still needs to be copied into a new buffer

Reply:
`temp_data` is a `device_scalar<string_view>`. The `string_scalar` will allocate `device_buffer` in GPU (which it owns) based on input `string_view` size. `device_scalar<string_view>` will not be owned, it's content is transferred to CPU and used for allocation, then data is copied to the newly created `device_buffer`.